File: src/5_tmle/tmle.py
import pandas as pd
import numpy as np
import os

from tmle import tmle
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.neural_network import MLPClassifier
from xgboost import XGBClassifier


# Define the SL library
SL_library = [
    LogisticRegression(),
    RandomForestClassifier(),
    MLPClassifier(),
    XGBClassifier()
]


setting = "sens/xgb_cv_all_coh"

### Constants ###
# Number of folds used in cross-validation (also used as parallel processes)
N_FOLDS = 5
# Tests per cohort
NREP = 50

### Get the data ###
# now read treatment from txt
with open("config/treatments.txt", "r") as f:
    treatments = f.read().splitlines()
treatments.remove("treatment")

# read features from list in txt
with open("config/confounders.txt", "r") as f:
    confounders = f.read().splitlines()
confounders.remove("confounder")

# read the cofounders from list in txt
with open("config/outcomes.txt", "r") as f:
    outcomes = f.read().splitlines()
outcomes.remove("outcome")

# Get the cohorts
with open("config/cohorts.txt", "r") as f:
    cohorts = f.read().splitlines()
cohorts.remove("cohorts")

# Get cancer types:
with open("config/cancer_types.txt", "r") as f:
    cancer_types = f.read().splitlines()
cancer_types.remove("cancer_type")


def calculate_tmle(A, Y, W, SL_library, Q_SL_library=None):
    """
    Calculates the TMLE estimate for the average treatment effect (ATE).
    :param A: Treatment variable
    :param Y: Outcome variable
    :param W: Confounder variables
    :param SL_library: List of sklearn estimators for the Super Learner
    :param Q_SL_library: List of sklearn estimators for the Super Learner
    :return: TMLE estimate for the ATE
    """
    
    if not(Q_SL_library):
        Q_SL_library = SL_library

    # Define the TMLE model
    tmle_model = tmle(Y=Y, A=A, W=W, family="binomial", gbound=[0.05, 0.95], g_SL_library=SL_library, Q_SL_library=SL_library)

    # Run TMLE
    result = tmle_model.run()

    # Access the results
    print("ATE Estimate:", result.ATE)
    print("ATE Confidence Interval:", result.ATE_confidence_interval)
    print("p-value:", result.p_value)

    print('Result')

    return result


def calculate_tmle_per_cohort(data, groups, treatments, outcomes, confounders, cohort, results_df):
    for outcome in outcomes:
        # Get the treatments:
        for treatment in treatments:
            print(f"Doing the prediction for treatment: {treatment}")

            for group in groups:
                print(f"Group: {group}")
                data = data[data[group] == 1].drop(group, axis=1)

                # append treatments that are not the current one to confounders
                # select X, y
                conf = confounders + [t for t in treatments if t != treatment] 

                # compute OR based on all data
                W = data[conf]
                A = data[treatment]
                Y = data[outcome]

                results = calculate_tmle(A, Y, W, SL_library, Q_SL_library=None)

                results = {
                    'outcome': [outcome],
                    'treatment': [treatment],
                    'cohort': [group],
                    'race':np.nan,
                    # Add sev_min and sev_max
                    'prob_mort_start':np.nan,
                    'prob_mort_end':np.nan,
                    'psi': results.ATE.psi,
                    'i_ci': results.ATE.CI[0],
                    's_ci': results.ATE.CI[1],
                    'pvalue': results.ATE.pvalue,
                    'n': len(data),
                    'SL_libraries': " ".join(SL_library),
                    'Q_weights': " ".join(results.Qinit.coef),
                    'g_weights': " ".join(results.g.coef)
                }

                results = pd.DataFrame.from_dict(results)
                
                # append results to dataframe
                results_df = pd.concat([results_df, results], ignore_index=True)

        return results_df



def check_columns_in_df(df, columns):
    cols_not_in_df = []
    for col in columns:
        if col not in df.columns:
            cols_not_in_df.append(col)
            print(f"Column {col} not in df")
    if len(cols_not_in_df) > 0:
        print(f"This cofounders are not in the df: {cols_not_in_df}")
        return False
    else:
        return True


# create dataframes to store results
results_template = pd.DataFrame(columns=["outcome",
                                "treatment",
                                "cohort",
                                "race",
                                "prob_mort_start",
                                "prob_mort_end",
                                "psi",
                                "i_ci",
                                "s_ci",
                                "pvalue",
                                "n",
                                "SL_libraries",
                                "Q_weights",
                                "g_weights"])
results_df = results_template.copy()

group = ''
for cohort in cohorts:
    # Get the cohort data
    if cohort == 'cancer_vs_nocancer':
        # Get all data
        df = pd.read_csv(f"data/cohorts/merged_all.csv")
        group = 'has_cancer'
        cohort = 'cancer'

        """ Get provisional cofounders from the dataframe using the dtypes and excluding the treatments """
        confounders = [col for col in df.columns if df[col].dtype in ['float64', 'int64'] and col not in treatments]
        print(f"Confounders: {confounders}")
        check = check_columns_in_df(df, confounders)
        if check == False:
            continue

        results_df_aux = calculate_tmle_per_cohort(df, [group], treatments, confounders, outcomes, cohort+'_vs_others', results_template)
        results_df = pd.concat([results_df, results_df_aux], ignore_index=True)

    elif cohort == 'cancer_type':
        # Get the dataset for each cancer type
        for cancer_type in cancer_types:
            group = cancer_type
            print(f"Getting data for cancer type: {cancer_type}")
            df = pd.read_csv(f"data/cohorts/merged_cancer.csv")
            cohort = cancer_type

            """ Get provisional cofounders from the dataframe using the dtypes and excluding the treatments """
            confounders = [col for col in df.columns if df[col].dtype in ['float64', 'int64'] and col not in treatments]
            print(f"Confounders: {confounders}")
            check = check_columns_in_df(df, confounders)
            if check == False:
                continue
                
            results_df_aux = calculate_tmle_per_cohort(df, [group], treatments, confounders, outcomes, cohort+'_vs_others', results_template)
            results_df = pd.concat([results_df, results_df_aux], ignore_index=True)

    else:
        print(f"Error: {cohort} should be cancer_vs_nocancer or cancer_type or both of them")
        continue

# save results as we go
try:
    results_df.to_csv(f"results/models/{setting}.csv", index=False)
# if folder does not exist, create it and save results
except:
    try:
        os.mkdir("results/models")
        results_df.to_csv(f"results/models/{setting}.csv", index=False)
    except:
        # setting contains a slash, so we need to create the folder
        os.mkdir(f"results/models/{setting.split('/')[0]}")
        results_df.to_csv(f"results/models/{setting}.csv", index=False)
        

File: src/5_tmle/tmle.R
library(tmle)
library(pROC)
library(data.table)

### Get the data ###
# now read treatment from txt
treatments <- read.delim("config/treatments.txt")$treatment

# read features from list in txt
confounders <- read.delim("config/confounders.txt")$confounder

# read the cofounders from list in txt
outcomes <- read.delim("config/outcomes.txt")$outcome

# Get the cohorts
cohorts <- read.delim("config/cohorts.txt")$cohorts

# Get cancer types:
cancer_types <- read.delim("config/cancer_types.txt")$cancer_type

# Define the SL library
SL_library <- read.delim("config/SL_libraries_base.txt")

# Define predicted mortality ranges
prob_mort_ranges <- read.csv("config/prob_mort_ranges.csv")

# First iteration
FIRST <- TRUE

# run TMLE 
run_tmle <- function(data, treatment, confounders, outcome, SL_libraries,
                     cohort, sev_min, sev_max, results_df, group_true) {
               
    W <- data[, confounders]
    A <- data[, treatment]
    Y <- data[, outcome]

    if (length(unique(Y)) > 2) {
        
        # Normalize continuous outcomes to be between 0 and 1
        min.Y <- min(Y)
        max.Y <- max(Y)
        Y <- (Y-min.Y)/(max.Y-min.Y)
        # Transform continuous outcomes to be between 0 and 1
        Y <- (data[, outcome]-min.Y)/(max.Y-min.Y)

        result <- tmle(
            Y = Y,
            A = A,
            W = W,
            #Delta = my_delta,
            family = "gaussian", 
            gbound = c(0.05, 0.95),
            g.SL.library = SL_libraries$SL_library,
            Q.SL.library = SL_libraries$SL_library
            )

    log <- summary(result) 
    
    RR <- 0

    # Transform back the ATE estimate
    log$estimates$ATE$psi <- (max.Y-min.Y)*log$estimates$ATE$psi

    # Transform back the CI estimate
    log$estimates$ATE$CI[1] <- (max.Y-min.Y)*log$estimates$ATE$CI[1]
    log$estimates$ATE$CI[2] <- (max.Y-min.Y)*log$estimates$ATE$CI[2]

    }

    else {
        result <- tmle(
                    Y = Y,
                    A = A,
                    W = W,
                    #Delta = my_delta,
                    family = "binomial", 
                    gbound = c(0.05, 0.95),
                    g.SL.library = SL_libraries$SL_library,
                    Q.SL.library = SL_libraries$SL_library
                    )
    log <- summary(result)
    
    RR <- log$estimates$RR$psi # output RR for calculating the e-value 
    }

    print('***************')
   #print(log)

    results_df[nrow(results_df) + 1,] <- c( outcome,
                                            treatment,
                                            cohort,
                                            group_true,
                                            sev_min,
                                            sev_max,
                                            log$estimates$ATE$psi,
                                            log$estimates$ATE$CI[1],
                                            log$estimates$ATE$CI[2],
                                            log$estimates$ATE$pvalue,
                                            nrow(data),
                                            paste(SL_libraries$SL_library, collapse = " "),
                                            paste(result$Qinit$coef, collapse = " "),
                                            paste(result$g$coef, collapse = " "),
                                            RR                                    
                                            ) 
    return (results_df)
}


calculate_tmle_per_cohort <- function(data, groups, treatments, outcomes, confounders, cohort, results_df, SL_libraries) {

    for (outcome in outcomes) {
        print('***************')
        cat(paste("Outcomes:", outcome), "\n")
        # Get the treatments:
        for (treatment in treatments) {
            cat(paste("Doing the prediction for treatment:", treatment), "\n")
        
            for (group in groups) {
                cat(paste("Group:", group), "\n")

                # append treatments that are not the current one to confounders
                conf <- c()
                for (confounder in confounders) {
                    if (confounder != treatment) {
                        # Append treatment to confounders
                        conf <- c(conf, confounder)
                    }
                }

                # Get the data for the current group
                # When group is true: group = 1
                group_true = 1
                data_subset <- subset(data, data[[group]] == 1)

                if (group %in% c("group_solid", "group_hematologic", "group_metastasized")) {
                        
                        sev_min <- 0
                        sev_max <- 1

                        print(paste0("Stratification by prob_mort: ", sev_min, " - ", sev_max))

                        # Stratify by prob_mort
                        data_subsub <- subset(data_subset, prob_mort >= sev_min & prob_mort < sev_max)

                        # Run TMLE
                        results_df = run_tmle(data_subsub, treatment, confounders, outcome, SL_libraries,
                                                cohort, sev_min, sev_max, results_df, group_true)

                    } else {

                        for (i in 1:nrow(prob_mort_ranges)) {

                        sev_min <- prob_mort_ranges$min[i]
                        sev_max <- prob_mort_ranges$max[i]
                    
                        print(paste0("Stratification by prob_mort: ", sev_min, " - ", sev_max))

                        # Stratify by prob_mort
                        data_subsub <- subset(data_subset, prob_mort >= sev_min & prob_mort < sev_max)
                        
                        # Run TMLE                        
                        results_df = run_tmle(data_subsub, treatment, confounders, outcome, SL_libraries,
                                            cohort, sev_min, sev_max, results_df, group_true)

                    }
                }

                # When group is false: group = 0
                if (FIRST == TRUE) {
                    
                    FIRST <- FALSE
                    
                    group_true = 0
                    data_subset <- subset(data, data[["has_cancer"]] == 0)
                    
                    if (group %in% c("group_solid", "group_hematologic", "group_metastasized")) {
                            
                            sev_min <- 0
                            sev_max <- 1

                            print(paste0("Stratification by prob_mort: ", sev_min, " - ", sev_max))

                            # Stratify by prob_mort
                            data_subsub <- subset(data_subset, prob_mort >= sev_min & prob_mort < sev_max)

                            # Run TMLE
                            results_df = run_tmle(data_subsub, treatment, confounders, outcome, SL_libraries,
                                                    cohort, sev_min, sev_max, results_df, group_true)
                    
                    } else {

                        for (i in 1:nrow(prob_mort_ranges)) {

                            sev_min <- prob_mort_ranges$min[i]
                            sev_max <- prob_mort_ranges$max[i]
                        
                            print(paste0("Stratification by prob_mort: ", sev_min, " - ", sev_max))

                            # Stratify by prob_mort
                            data_subsub <- subset(data_subset, prob_mort >= sev_min & prob_mort < sev_max)
                            
                            # Run TMLE                        
                            results_df = run_tmle(data_subsub, treatment, confounders, outcome, SL_libraries,
                                                cohort, sev_min, sev_max, results_df, group_true)

                            print(paste0("Stratification by prob_mort: ", sev_min, " - ", sev_max))

                            # Stratify by prob_mort
                            data_subsub <- subset(data_subset, prob_mort >= sev_min & prob_mort < sev_max)
                        
                            # Run TMLE
                            results_df = run_tmle(data_subsub, treatment, confounders, outcome, SL_libraries,
                                                    cohort, sev_min, sev_max, results_df, group_true)
                        }
                    }
                }   
            }
        }
    }
    return(results_df)
}

check_columns_in_df <- function(df, columns) {
  cols_not_in_df <- setdiff(columns, colnames(df))
  
  if (length(cols_not_in_df) > 0) {
    cat("These cofounders are not in the df:", cols_not_in_df, "\n")
    return(FALSE)
  } else {
    return(TRUE)
  }
}

databases = c("all", "eicu", "mimic") 

for (db in databases){
  print('***************')
  print(db)
  print('***************')
  
    # create data.frames to store results
    results_df <- data.frame(matrix(ncol=15, nrow=0))
    colnames(results_df) <- c(
                            "outcome",
                            "treatment",
                            "cohort",
                            "has_cancer",
                            "prob_mort_start",
                            "prob_mort_end",
                            "ATE",
                            "i_ci",
                            "s_ci",
                            "pvalue",
                            "n",
                            "SL_libraries",
                            "Q_weights",
                            "g_weights",
                            "RR"
                           )
                        
    group <- ""
    for (cohort in cohorts) {
        
        # Remove hospital confounders for analysis with MIMIC only, as they do not vary
        confounders_aux <- confounders[!confounders %in% c("hospitalid", "numbedscategory", "teaching_hospital", "region")]

        if (cohort == "cancer_vs_nocancer") {

            df <- read.csv("data/cohorts/merged_all.csv")

            # Get all data
            if (db == "eicu") {
               df <- subset(df, source == db)

            } else if (db == "mimic") {
               df <- subset(df, source == db)

            } else {
                cat(paste("Error:", db, "should be all, eicu or mimic"), "\n")
                break
            }

            group <- "has_cancer"
            cohort <- "cancer"
            
            if (db == "mimic") {
            # pass updated confounders without hospital confounders to tmle function
            results_df <- calculate_tmle_per_cohort(df, group, treatments, outcomes, confounders_aux, paste0(cohort, "_vs_others"), results_df, SL_library)
            
            } else {
            # if database is all or eicu -> keep hospital confounders
            results_df <- calculate_tmle_per_cohort(df, group, treatments, outcomes, confounders, paste0(cohort, "_vs_others"), results_df, SL_library)
            } 
        }
        else if (cohort == "cancer_type") {
            
            # Remove hospital confounders for analysis with MIMIC only, as they do not vary
            confounders_aux <- confounders[!confounders %in% c("hospitalid", "numbedscategory", "teaching_hospital", "region")]

            for (cancer_type in cancer_types) {
                cat(paste("Getting data for cancer type:", cancer_type), "\n")
    
                # Get all data
                if (db == "all"){
                    df <- read.csv("data/cohorts/merged_all.csv")
                } else if (db == "eicu") {
                df <- read.csv("data/cohorts/merged_eicu_all.csv")
                } else if (db == "mimic") {
                df <- read.csv("data/cohorts/merged_mimic_all.csv")
                } else {
                    cat(paste("Error:", db, "should be all, eicu or mimic"), "\n")
                    break
                }

                group <- cancer_type
                cohort <- cancer_type
                
                if (db == "mimic") {
                # pass updated confounders without hospital confounders to tmle function
                results_df <- calculate_tmle_per_cohort(df, group, treatments, outcomes, confounders_aux, paste0(cohort, "_vs_nocancer"), results_df, SL_library)
            
                } else {
                # if database is all or eicu -> keep hospital confounders
                results_df <- calculate_tmle_per_cohort(df, group, treatments, outcomes, confounders, paste0(cohort, "_vs_others"), results_df, SL_library)
                }
            }     
        } else {
            cat(paste("Error:", cohort, "should be cancer_vs_nocancer or cancer_type or both of them"), "\n")
            next
        }
    }

    # Save results as we go
    dir.create("results/tmle", showWarnings = FALSE, recursive = TRUE)
    results <- paste0('tmle_results_', db)
    write.csv(results_df, file.path("results/tmle", paste0(results, ".csv")), row.names = FALSE)

}
